### PR TITLE
Create data retrieval API for the Academix Project  

### PR DESCRIPTION
**Pull request by Gravewalker666**
## Purpose
This is a part of #35 

## Goals
To Create necessary endpoints to retrieve data

## Approach
- Created repositories for item, sub-category and, category
- Created projections for each entity
- Created academic service and controller

## Endpoints and Payloads
- Categories
image
- SubCategories by Category
image
- Items by SubCategory
image

##  Checklist
- [x] This PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.
- [x] I have read and understood the development best practices guidelines ( http://bit.ly/sef-best-practices )
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented on my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation

## Related PRs
 
#36 

## Test environment
- OpenJDK version "1.8.0_252"
- Tomcat 9.0.35
- Postman
- Ubuntu 19.10

## Learning
- https://vladmihalcea.com/dto-projection-jpa-query/
- https://www.baeldung.com/jpa-join-types
- https://stackoverflow.com/questions/53322203/spring-boot-data-query-to-dto
- https://stackoverflow.com/questions/52915532/return-a-dto-object-in-spring-data-repository-method-annotated-with-query/52919487
- https://www.baeldung.com/jpa-queries
- https://www.bytestree.com/spring/spring-data-jpa-projections-5-ways-return-custom-object/
- https://www.baeldung.com/spring-data-rest-projections-excerpts
- https://thorben-janssen.com/projections-with-jpa-and-hibernate/
- https://docs.spring.io/spring-data/jpa/docs/current/reference/html/#projections
- https://javadeveloperzone.com/spring/spring-jpa-projection-example/

### CODE REVIEW

File: backend/src/main/java/org/sefglobal/core/projections/CustomItem.java [hunk 1]
```diff
@@ -0,0 +1,9 @@
+package org.sefglobal.core.projections;
+
+import java.util.Set;
+
+public interface CustomItem {
+    Long getId();
+    String getLink();
+    Set<CustomTranslation> getTranslations();
+}
```
Comment (YohanAvishke) [inline backend/src/main/java/org/sefglobal/core/projections/CustomItem.java:9]: Include item's descriptions

**Comment by YohanAvishke** [comment]
Change `category` projection to match [payload](http://www.mocky.io/v2/5ecf3fa63200002400e3cffb)

File: backend/src/main/java/org/sefglobal/core/academix/controller/ItemController.java [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    private final
+    ItemService itemService;
+
+    public ItemController(ItemService itemService) {
+        this.itemService = itemService;
+    }
+
+    @GetMapping("/academix/sub-categories/{id}/items")
+    @ResponseStatus(HttpStatus.OK)
+    public List<CustomItem> getAllItemsBySubCategoryId(@PathVariable(value = "id") Long id,
```
Comment (YohanAvishke) [inline backend/src/main/java/org/sefglobal/core/academix/controller/ItemController.java:22]:
```suggestion
    public List<CustomItem> getAllItemsBySubCategoryId(@PathVariable Long id,
```

File: backend/src/main/java/org/sefglobal/core/academix/controller/ItemController.java [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+import org.sefglobal.core.academix.service.ItemService;
+import org.springframework.http.HttpStatus;
+import org.springframework.web.bind.annotation.*;
+
+import java.util.List;
+
+@RestController
+public class ItemController {
+    private final
+    ItemService itemService;
```
Comment (YohanAvishke) [inline backend/src/main/java/org/sefglobal/core/academix/controller/ItemController.java:14]: Remove line breaker

File: backend/src/main/java/org/sefglobal/core/academix/controller/SubCategoryController.java [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+import org.springframework.web.bind.annotation.ResponseStatus;
+import org.springframework.web.bind.annotation.RestController;
+
+import java.util.List;
+
+@RestController
+public class SubCategoryController {
+
+    private final
+    SubCategoryService subCategoryService;
```
Comment (YohanAvishke) [inline backend/src/main/java/org/sefglobal/core/academix/controller/SubCategoryController.java:18]: Remove line breaker

**Comment by YohanAvishke** [comment]
Hi, is there any reason for handling `getAllItemsBySubCategory ` in `ItemService ` and `getAllSubCategoryByCategory ` in `SubCategoryService ` in two different approaches?

File: backend/src/main/java/org/sefglobal/core/academix/service/ItemService.java [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+        this.subCategoryRepository = subCategoryRepository;
+        this.itemRepository = itemRepository;
+    }
+
+    public List<CustomItem> getAllItemsBySubCategory(Long subCategoryId, int pageNumber, int pageSize) throws BadRequestException {
+        Pageable pageable = PageRequest.of(pageNumber, pageSize);
+        SubCategory subCategory = subCategoryRepository.findById(subCategoryId)
+                .orElseThrow(() -> new BadRequestException("Error, Sub Category by id:"+subCategoryId+" doesn't exist."));
+        return itemRepository.getAllBySubCategories(subCategory, pageable);
+    }
```
Comment (YohanAvishke) [inline backend/src/main/java/org/sefglobal/core/academix/service/ItemService.java:29]: Hi, is there any reason for handling this method and `getAllSubCategoryByCategory ` method in `SubCategoryService ` in two different approaches?

File: backend/src/main/java/org/sefglobal/core/academix/service/ItemService.java [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+public class ItemService {
+    private final SubCategoryRepository subCategoryRepository;
+    private final ItemRepository itemRepository;
+
+    public ItemService(SubCategoryRepository subCategoryRepository, ItemRepository itemRepository) {
+        this.subCategoryRepository = subCategoryRepository;
+        this.itemRepository = itemRepository;
+    }
+
+    public List<CustomItem> getAllItemsBySubCategory(Long subCategoryId, int pageNumber, int pageSize) throws BadRequestException {
```
Comment (YohanAvishke) [inline backend/src/main/java/org/sefglobal/core/academix/service/ItemService.java:24]: Return a `Page`

File: backend/src/main/java/org/sefglobal/core/academix/controller/ItemController.java (hunk 2 above)
Comment (YohanAvishke) [inline backend/src/main/java/org/sefglobal/core/academix/controller/ItemController.java:22]: Return a `Page`

File: backend/src/main/java/org/sefglobal/core/academix/repository/ItemRepository.java [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+import org.sefglobal.core.academix.model.Item;
+import org.sefglobal.core.academix.model.SubCategory;
+import org.sefglobal.core.academix.projections.CustomItem;
+import org.springframework.data.domain.Pageable;
+import org.springframework.data.jpa.repository.JpaRepository;
+
+import java.util.List;
+
+public interface ItemRepository extends JpaRepository<Item, Long> {
+    List<CustomItem> getAllBySubCategories(SubCategory subCategory, Pageable pageable);
```
Comment (YohanAvishke) [inline backend/src/main/java/org/sefglobal/core/academix/repository/ItemRepository.java:12]: Return a `Page`

File: backend/src/main/java/org/sefglobal/core/academix/repository/CategoryRepository.java [hunk 8]
```diff
@@ -0,0 +1,11 @@
+package org.sefglobal.core.academix.repository;
+
+import org.sefglobal.core.academix.model.Category;
+import org.sefglobal.core.academix.projections.CustomCategory;
+import org.springframework.data.jpa.repository.JpaRepository;
+
+import java.util.List;
+
+public interface CategoryRepository extends JpaRepository<Category, Long> {
+    List<CustomCategory> findAllBy();
```
Comment (YohanAvishke) [inline backend/src/main/java/org/sefglobal/core/academix/repository/CategoryRepository.java:10]: A method already exists in `PagingAndSortingRepository` to `findAll` use that one.

**Comment by YohanAvishke** [comment]
Please add Java doc comments to all your service methods.

File: backend/src/main/java/org/sefglobal/core/academix/repository/CategoryRepository.java (hunk 8 above)
Comment (Gravewalker666) [reply backend/src/main/java/org/sefglobal/core/academix/repository/CategoryRepository.java:10]: We can't use the projection if we use the existing `findAll` method.

Comment (YohanAvishke) [reply backend/src/main/java/org/sefglobal/core/academix/repository/CategoryRepository.java:10]: Have a look into Spring `Excerpts`, you should be able to achieve this by using `Excerpts`

Comment (Gravewalker666) [reply backend/src/main/java/org/sefglobal/core/academix/repository/CategoryRepository.java:10]: We cant use excerpts if we are using `@RestController`, to use `excerpts` we have to add `@RepositoryRestResource` annotation in our repository and it's incompatible with custom written `@RestController` endpoints.
- https://stackoverflow.com/questions/34393874/spring-boot-restcontroller-endpoints-and-spring-data-rest-projections
- https://stackoverflow.com/questions/33172317/spring-boot-projection-with-restcontroller

**Pull request merged**